### PR TITLE
Implementing exclude paths for FIM

### PR DESCRIPTION
**Pull request by uptycs-nishant**
Added exclude path support - (files/directories) for which events not to be propagated.
  "exclude_paths": {
    "homes": [
      "/home/not_to_monitor/.ssh/%%"
    ],
    "tmp": [
      "/tmp/too_many_events/"
    ]
 }

### CODE REVIEW

**Comment by theopolis** [comment]
Can you include some test cases too? INotity had unit and end2end testing so it should be semi-straightforward.

File: osquery/events/pathset.h [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+ * This PathSet can take any of the two policies -
+ * 1. patternedPath - path can contain pattern '%' and '%%'.
+ *
+ * 2. resolvedPath - path is resolved before being inserted into set.
+ *                   But path can match recursively.
+ *
+ */
+
+template <typename PathType>
+class PathSet {
```
Comment (theopolis) [inline osquery/events/pathset.h]: This should be non-copyable (using boost's noncopyable).

**Comment by theopolis** [comment]
Thanks for this! The last thing this needs is tests for FSEvents, then it's ready to go!

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 978709ed82211afc742c1ce5d3afb1de1b738ecf (Job results: [1594](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1594/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 978709ed82211afc742c1ce5d3afb1de1b738ecf (Job results: [1595](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1595/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit c24e37164920e4db9e5d1a0edb78f4647743cafa (Job results: [1596](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1596/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit c24e37164920e4db9e5d1a0edb78f4647743cafa (Job results: [1597](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1597/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 90f5a0a729f9bb821373462c1c0d6647f5dc8dd6 (Job results: [1600](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1600/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 90f5a0a729f9bb821373462c1c0d6647f5dc8dd6 (Job results: [1601](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1601/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 5e9e5c0d51a96042f40c48b8057b528fa948e8fd (Job results: [1614](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1614/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 5e9e5c0d51a96042f40c48b8057b528fa948e8fd (Job results: [1615](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1615/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit b937725e0b5a3817da74293c469dba3bef4417c4 (Job results: [1616](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1616/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit b937725e0b5a3817da74293c469dba3bef4417c4 (Job results: [5507](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5507/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit b937725e0b5a3817da74293c469dba3bef4417c4 (Job results: [1617](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1617/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 4f0978a66e23f44c96f686eba3d50292f1393f71 (Job results: [1618](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1618/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 4f0978a66e23f44c96f686eba3d50292f1393f71 (Job results: [1619](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1619/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 6d12f141b71a2fb5edab3f19d763d664515ddca1 (Job results: [509](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/509/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 6d12f141b71a2fb5edab3f19d763d664515ddca1 (Job results: [510](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/510/)) failed one or more tests (**FreeBSD**).

**Comment by theopolis** [comment]
This looks great! Thanks for the work @uptycs-nishant. I may want to tweak a few things later on, and I'll want to stress a few situations: (1) where an exclude path is 90% of the potential subscriber events, (2) when mount events occur for FSEvents, (3) dotfiles in the base of directories set in the list of monitored paths.

I think some re-organization of filesystem utilities is due too. But we can take care of that in subsequent pull requests.

**Pull request merged**